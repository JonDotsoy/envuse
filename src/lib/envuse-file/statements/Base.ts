import { BufferCursor } from "./BufferCursor";
import { Iter } from "./Iter";
import { UnexpectedTokenError } from "./UnexpectedTokenError";

type B<T> = T extends { propsMutable: infer R } ? R extends keyof T ? Partial<Pick<T, R>> : {} : {}

export abstract class Base {
  end!: number;
  children: Base[] = [];

  raw = Buffer.from([]);
  _raw: string = '';
  #filename: string | null;
  #bufferCursor: BufferCursor
  #body: Buffer

  constructor(
    filename: string | null,
    body: Buffer,
    public pos: number,
    bufferCursor = new BufferCursor(body),
  ) {
    this.#filename = filename;
    this.#bufferCursor = bufferCursor
    this.#body = body
  }

  get filename() { return this.#filename; }
  get bufferCursor() { return this.#bufferCursor; }
  get body() { return this.#body; }

  private load() {
    this.prepare(this.bufferCursor);
    return this;
  }

  createElement<T extends Base>(Comp: { new(filename: string | null, body: Buffer, pos: number, bufferCursor?: BufferCursor): T }, assign?: B<T>) {
    const comp = new Comp(this.filename, this.body, this.bufferCursor.position, this.bufferCursor);

    return Base.createElement(comp, assign)
  }

  static createElement<T extends Base>(comp: T, assign?: B<T>) {
    if (assign) {
      Object.assign(comp, assign)
    }

    return comp.load()
  }

  abstract prepare(bufferCursor: BufferCursor): void;

  // f(cb: (gen: Iter, ) => void) {
  //   const gen = this.ngen();envuseFileParser.toAstBody()
  //   while (true) {
  //     const { done, value } = gen.next();
  //     if (done || !value) return;
  //     const [index, current_char, { prev, next }] = value;

  //     cb
  //   }
  // }

  appendRaw(raw: Buffer | number) {
    this.raw = Buffer.concat([this.raw, typeof raw === 'number' ? Buffer.from([raw]) : raw]);
    this._raw = this.raw.toString();

    return this;
  }

  *iter() {
    let index = this.pos;

    while (true) {
      const char = this.body[index];
      if (char === undefined)
        break;

      const current_index = index;
      const current_char = Buffer.from([char])
      const prev = (len: number) => {
        const begin = Math.max(index - len + 1, 0);
        const end = Math.min(index + 1, this.body.length);
        return this.body.slice(begin, end);
      };
      const next = (len: number) => {
        const begin = Math.max(index, 0);
        const end = Math.min(index + len, this.body.length);
        return this.body.slice(begin, end);
      };

      yield [index, current_char, { current_char, current_index, prev, next }] as const;

      index += 1;
    }
  }

  rejectUnexpectedTokenError(): never {
    const err = new UnexpectedTokenError(this.filename, this.body, this.bufferCursor.position)

    Error.captureStackTrace(err, Base.prototype.rejectUnexpectedTokenError)

    throw err;
  }

  toJSON() {
    return {}
  }
}
