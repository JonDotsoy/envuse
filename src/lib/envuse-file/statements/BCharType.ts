
export type BCharType = number | undefined;
