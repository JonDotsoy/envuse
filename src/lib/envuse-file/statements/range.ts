export function* range(from: number, to: number) {
  if (from > to)
    throw new Error('param from is greater than param to');
  for (let current = from; current <= to; current += 1) {
    yield current;
  }
}
