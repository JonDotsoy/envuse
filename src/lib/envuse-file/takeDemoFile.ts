import path from 'path';
import fs from 'fs';

const mem: { [k: string]: number } = {}
const getCount = (name: string) => {
  mem[name] = (mem[name] ?? 0) + 1
  return mem[name];
}

export function takeDemoFile(ext?: string) {
  const testPath = expect.getState().testPath;
  const filedemo = `${path.dirname(testPath)}/__demo__/${path.basename(testPath)}.${getCount(testPath)}.demo${ext ?? ''}`;
  fs.mkdirSync(path.dirname(filedemo), { recursive: true });

  if (!fs.existsSync(filedemo)) {
    fs.writeFileSync(filedemo, '')
    console.log('+ file demo created')
    return [filedemo, Buffer.from([])] as const
  } else {
    return [filedemo, fs.readFileSync(filedemo)] as const
  }
}
