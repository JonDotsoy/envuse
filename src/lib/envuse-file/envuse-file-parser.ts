import { Base } from "./statements/Base";
import { Block } from "./statements/Root";
import { UnexpectedTokenError } from "./statements/UnexpectedTokenError";

/** AST Parser */
export class EnvuseFileParser {
  constructor(
    private filename: string,
    private body: Buffer
  ) { }

  toAstBody() {
    try {
      return Base.createElement(new Block(this.filename, this.body, 0));
    } catch (err) {
      if (UnexpectedTokenError.isUnexpectedTokenError(err)) {
        // Error.captureStackTrace(err, EnvuseFileParser.prototype.toAstBody);
        // const stack = err.stack;
        // if (stack) {
        //   // let posEndMessage = err.name.length + err.message.length + 3;
        // }
      }
      throw err;
    }
  }
}