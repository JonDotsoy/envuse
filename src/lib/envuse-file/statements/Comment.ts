import { Base } from "./Base";
import { BufferCursor } from "./BufferCursor";
import { BCharType } from "./BCharType";


export class Comment extends Base {
  prepare(bufferCursor: BufferCursor<BCharType>): void {
    while (bufferCursor.has()) {
      if (bufferCursor.current() === 0x0a) {
        this.end = bufferCursor.position
        bufferCursor.forward();
        return;
      }

      this.appendRaw(bufferCursor.current());
      bufferCursor.forward();
    }
  }
}
