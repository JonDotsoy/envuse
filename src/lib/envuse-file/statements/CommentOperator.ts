import { Base } from "./Base";
import { BufferCursor } from "./BufferCursor";
import { BCharType } from "./BCharType";
import { Space } from "./Space";
import { VariableKey } from "./Variable";
import { Block } from "./Root";


export class CommentOperatorStatement extends Base {
  prepare(bufferCursor: BufferCursor<BCharType>): void {
    while (bufferCursor.has()) {
      if (bufferCursor.current() === 0x0a) {
        this.end = bufferCursor.position;
        bufferCursor.forward();
        return;
      }

      this.appendRaw(bufferCursor.current());
      bufferCursor.forward();
    }
  }
}


export class CommentOperator extends Base {
  operator!: VariableKey;
  statement!: CommentOperatorStatement;
  block!: Block;

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    bufferCursor.forward(2)

    if (bufferCursor.current() === 0x20) this.createElement(Space)

    this.operator = this.createElement(VariableKey)

    switch (this.operator.raw.toString()) {
      case 'if': break;
      case 'fi': break;
      default: {
        bufferCursor.backward(this.operator.raw.length);
        this.rejectUnexpectedTokenError();
      }
    }

    if (this.operator.raw.equals(Buffer.from('fi'))) {
      return;
    }

    this.createElement(Space)

    this.statement = this.createElement(CommentOperatorStatement)

    this.block = this.createElement(Block, {
      handleCheckCloseBlock() {
        const lastChild = this.children[this.children.length - 1]
        if (lastChild instanceof CommentOperator && lastChild.operator.raw.equals(Buffer.from('fi'))) {
          return true;
        }
        return false;
      }
    })
  }
}
