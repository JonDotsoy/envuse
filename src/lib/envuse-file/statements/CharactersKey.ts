import { range } from "./range";

export namespace CharactersKey {
  /** symbol: . */
  export const dot = 0x2e;
  /** symbol: # */
  export const numberSign = 0x23;
  /** symbols: 0123456789 */
  export const numbers = Array.from(range(0x30, 0x39));
  /** symbols: abcdefghijklmnopqrstuvwxyz */
  export const english_alphabet_lower = Array.from(range(0x61, 0x7a));
  /** symbols: ABCDEFGHIJKLMNOPQRSTUVWXYZ */
  export const english_alphabet_upper = Array.from(range(0x41, 0x5a));
  /** symbol: _ */
  export const underscore = 0x5f;
  /** symbol: - */
  export const hyphenMinus = 0x2d;
  /** symbol: " " */
  export const space = 0x20;
  /** symbol: "\n" */
  export const newLineLF = 0x0a;
  /** symbol: "\r" */
  export const carriageReturn = 0x0d;
  /** symbols: "\r\n" */
  export const newLineCRLF = [carriageReturn, newLineLF];
}
