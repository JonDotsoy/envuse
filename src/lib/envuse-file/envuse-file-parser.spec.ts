import { EnvuseFileParser } from "./envuse-file-parser";
import { takeDemoFile } from "./statements/takeDemoFile";
import util from 'util'

describe('EnvuseFileParser2', () => {

  it('shoud make a ast out', () => {
    const [filename1, body1] = takeDemoFile('.env');

    const envuseFileParser = new EnvuseFileParser(filename1, body1);
    expect(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody())).toMatchSnapshot()
  })

  it('shoud make a ast with operators', () => {
    const [filename, body] = takeDemoFile('.env');

    const envuseFileParser = new EnvuseFileParser(filename, body);
    expect(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody())).toMatchSnapshot()
    // console.log(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody()))
  })

  it('shoud make a ast with multiple blocks', () => {
    const [filename, body] = takeDemoFile('.env');

    const envuseFileParser = new EnvuseFileParser(filename, body);
    expect(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody())).toMatchSnapshot()
    // console.log(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody()))
  })

})
