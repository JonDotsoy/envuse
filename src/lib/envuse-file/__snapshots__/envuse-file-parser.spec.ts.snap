// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvuseFileParser2 shoud make a ast out 1`] = `
" Block {
  pos: 0,
  children: [
    SpaceNewLine {
      pos: 0,
      children: [],
      raw: <Buffer 20>,
      _raw: ' ',
      end: 1
    },
    Variable {
      pos: 1,
      children: [
        VariableKey {
          pos: 1,
          children: [],
          raw: <Buffer 61 62 63>,
          _raw: 'abc',
          end: 4
        },
        SymboEqual {
          pos: 4,
          children: [],
          raw: <Buffer 3d>,
          _raw: '=',
          end: 5
        },
        VariableValue {
          pos: 5,
          children: [],
          raw: <Buffer 64 65 66 20 23 20 61 20 63 6f 6d 6d 65 6e 74>,
          _raw: 'def # a comment',
          end: 20
        }
      ],
      raw: <Buffer >,
      _raw: '',
      keyVariable: VariableKey {
        pos: 1,
        children: [],
        raw: <Buffer 61 62 63>,
        _raw: 'abc',
        end: 4
      },
      valueVariable: VariableValue {
        pos: 5,
        children: [],
        raw: <Buffer 64 65 66 20 23 20 61 20 63 6f 6d 6d 65 6e 74>,
        _raw: 'def # a comment',
        end: 20
      }
    },
    Variable {
      pos: 21,
      children: [
        VariableKey {
          pos: 21,
          children: [],
          raw: <Buffer 66 6f 6f>,
          _raw: 'foo',
          end: 24
        },
        SymboEqual {
          pos: 24,
          children: [],
          raw: <Buffer 3d>,
          _raw: '=',
          end: 25
        },
        VariableValue {
          pos: 25,
          children: [],
          raw: <Buffer 62 61 72>,
          _raw: 'bar',
          end: 28
        }
      ],
      raw: <Buffer >,
      _raw: '',
      keyVariable: VariableKey {
        pos: 21,
        children: [],
        raw: <Buffer 66 6f 6f>,
        _raw: 'foo',
        end: 24
      },
      valueVariable: VariableValue {
        pos: 25,
        children: [],
        raw: <Buffer 62 61 72>,
        _raw: 'bar',
        end: 28
      }
    },
    SpaceNewLine {
      pos: 29,
      children: [],
      raw: <Buffer 0a 20 20 20 20 0a>,
      _raw: '\\\\n    \\\\n',
      end: 35
    },
    Variable {
      pos: 35,
      children: [
        VariableKey {
          pos: 35,
          children: [],
          raw: <Buffer 61 62 63>,
          _raw: 'abc',
          end: 38
        },
        Space {
          pos: 38,
          children: [],
          raw: <Buffer 20 20 20 20>,
          _raw: '    ',
          end: 42
        },
        SymboEqual {
          pos: 42,
          children: [],
          raw: <Buffer 3d>,
          _raw: '=',
          end: 43
        },
        VariableValue {
          pos: 43,
          children: [],
          raw: <Buffer 64 5c 22 65 66>,
          _raw: 'd\\\\\\\\\\"ef',
          end: 49
        }
      ],
      raw: <Buffer >,
      _raw: '',
      keyVariable: VariableKey {
        pos: 35,
        children: [],
        raw: <Buffer 61 62 63>,
        _raw: 'abc',
        end: 38
      },
      valueVariable: VariableValue {
        pos: 43,
        children: [],
        raw: <Buffer 64 5c 22 65 66>,
        _raw: 'd\\\\\\\\\\"ef',
        end: 49
      }
    },
    SpaceNewLine {
      pos: 50,
      children: [],
      raw: <Buffer 20>,
      _raw: ' ',
      end: 51
    },
    Comment {
      pos: 51,
      children: [],
      raw: <Buffer 23 20 63 6f 6d 65 6e 74 65 64>,
      _raw: '# comented',
      end: 61
    }
  ],
  raw: <Buffer >,
  _raw: '',
  end: 62
}"
`;

exports[`EnvuseFileParser2 shoud make a ast with multiple blocks 1`] = `
" Block {
  pos: 0,
  children: [
    Variable {
      pos: 0,
      children: [
        VariableKey {
          pos: 0,
          children: [],
          raw: <Buffer 43>,
          _raw: 'C',
          end: 1
        },
        SymboEqual {
          pos: 1,
          children: [],
          raw: <Buffer 3d>,
          _raw: '=',
          end: 2
        },
        VariableValue {
          pos: 2,
          children: [],
          raw: <Buffer 43>,
          _raw: 'C',
          end: 3
        }
      ],
      raw: <Buffer >,
      _raw: '',
      keyVariable: VariableKey {
        pos: 0,
        children: [],
        raw: <Buffer 43>,
        _raw: 'C',
        end: 1
      },
      valueVariable: VariableValue {
        pos: 2,
        children: [],
        raw: <Buffer 43>,
        _raw: 'C',
        end: 3
      }
    },
    CommentOperator {
      pos: 4,
      children: [],
      raw: <Buffer >,
      _raw: '',
      operator: VariableKey {
        pos: 6,
        children: [],
        raw: <Buffer 69 66>,
        _raw: 'if',
        end: 8
      },
      statement: CommentOperatorStatement {
        pos: 9,
        children: [],
        raw: <Buffer 74 72 75 65>,
        _raw: 'true',
        end: 13
      },
      block: Block {
        pos: 14,
        children: [
          Variable {
            pos: 14,
            children: [
              VariableKey {
                pos: 14,
                children: [],
                raw: <Buffer 61>,
                _raw: 'a',
                end: 15
              },
              SymboEqual {
                pos: 15,
                children: [],
                raw: <Buffer 3d>,
                _raw: '=',
                end: 16
              },
              VariableValue {
                pos: 16,
                children: [],
                raw: <Buffer 61>,
                _raw: 'a',
                end: 17
              }
            ],
            raw: <Buffer >,
            _raw: '',
            keyVariable: VariableKey {
              pos: 14,
              children: [],
              raw: <Buffer 61>,
              _raw: 'a',
              end: 15
            },
            valueVariable: VariableValue {
              pos: 16,
              children: [],
              raw: <Buffer 61>,
              _raw: 'a',
              end: 17
            }
          },
          CommentOperator {
            pos: 18,
            children: [],
            raw: <Buffer >,
            _raw: '',
            operator: VariableKey {
              pos: 20,
              children: [],
              raw: <Buffer 69 66>,
              _raw: 'if',
              end: 22
            },
            statement: CommentOperatorStatement {
              pos: 23,
              children: [],
              raw: <Buffer 66 61 6c 73 65>,
              _raw: 'false',
              end: 28
            },
            block: Block {
              pos: 29,
              children: [
                Variable {
                  pos: 29,
                  children: [
                    VariableKey {
                      pos: 29,
                      children: [],
                      raw: <Buffer 62>,
                      _raw: 'b',
                      end: 30
                    },
                    SymboEqual {
                      pos: 30,
                      children: [],
                      raw: <Buffer 3d>,
                      _raw: '=',
                      end: 31
                    },
                    VariableValue {
                      pos: 31,
                      children: [],
                      raw: <Buffer 62>,
                      _raw: 'b',
                      end: 32
                    }
                  ],
                  raw: <Buffer >,
                  _raw: '',
                  keyVariable: VariableKey {
                    pos: 29,
                    children: [],
                    raw: <Buffer 62>,
                    _raw: 'b',
                    end: 30
                  },
                  valueVariable: VariableValue {
                    pos: 31,
                    children: [],
                    raw: <Buffer 62>,
                    _raw: 'b',
                    end: 32
                  }
                },
                CommentOperator {
                  pos: 33,
                  children: [],
                  raw: <Buffer >,
                  _raw: '',
                  operator: VariableKey {
                    pos: 35,
                    children: [],
                    raw: <Buffer 66 69>,
                    _raw: 'fi',
                    end: 37
                  }
                }
              ],
              raw: <Buffer >,
              _raw: '',
              handleCheckCloseBlock: [Function: handleCheckCloseBlock],
              end: 37
            }
          },
          SpaceNewLine {
            pos: 37,
            children: [],
            raw: <Buffer 0a>,
            _raw: '\\\\n',
            end: 38
          },
          CommentOperator {
            pos: 38,
            children: [],
            raw: <Buffer >,
            _raw: '',
            operator: VariableKey {
              pos: 40,
              children: [],
              raw: <Buffer 66 69>,
              _raw: 'fi'
            }
          }
        ],
        raw: <Buffer >,
        _raw: '',
        handleCheckCloseBlock: [Function: handleCheckCloseBlock],
        end: 42
      }
    }
  ],
  raw: <Buffer >,
  _raw: '',
  end: 42
}"
`;

exports[`EnvuseFileParser2 shoud make a ast with operators 1`] = `
" Block {
  pos: 0,
  children: [
    Variable {
      pos: 0,
      children: [
        VariableKey {
          pos: 0,
          children: [],
          raw: <Buffer 46 4f 4f>,
          _raw: 'FOO',
          end: 3
        },
        SymboEqual {
          pos: 3,
          children: [],
          raw: <Buffer 3d>,
          _raw: '=',
          end: 4
        },
        VariableValue {
          pos: 4,
          children: [],
          raw: <Buffer 42 41 52>,
          _raw: 'BAR',
          end: 7
        }
      ],
      raw: <Buffer >,
      _raw: '',
      keyVariable: VariableKey {
        pos: 0,
        children: [],
        raw: <Buffer 46 4f 4f>,
        _raw: 'FOO',
        end: 3
      },
      valueVariable: VariableValue {
        pos: 4,
        children: [],
        raw: <Buffer 42 41 52>,
        _raw: 'BAR',
        end: 7
      }
    },
    CommentOperator {
      pos: 8,
      children: [],
      raw: <Buffer >,
      _raw: '',
      operator: VariableKey {
        pos: 11,
        children: [],
        raw: <Buffer 69 66>,
        _raw: 'if',
        end: 13
      },
      statement: CommentOperatorStatement {
        pos: 14,
        children: [],
        raw: <Buffer 74 72 75 65>,
        _raw: 'true',
        end: 18
      },
      block: Block {
        pos: 19,
        children: [
          Variable {
            pos: 19,
            children: [
              VariableKey {
                pos: 19,
                children: [],
                raw: <Buffer 61 61 61>,
                _raw: 'aaa',
                end: 22
              },
              SymboEqual {
                pos: 22,
                children: [],
                raw: <Buffer 3d>,
                _raw: '=',
                end: 23
              },
              VariableValue {
                pos: 23,
                children: [],
                raw: <Buffer 62 62 62>,
                _raw: 'bbb',
                end: 27
              }
            ],
            raw: <Buffer >,
            _raw: '',
            keyVariable: VariableKey {
              pos: 19,
              children: [],
              raw: <Buffer 61 61 61>,
              _raw: 'aaa',
              end: 22
            },
            valueVariable: VariableValue {
              pos: 23,
              children: [],
              raw: <Buffer 62 62 62>,
              _raw: 'bbb',
              end: 27
            }
          },
          SpaceNewLine {
            pos: 28,
            children: [],
            raw: <Buffer 0a>,
            _raw: '\\\\n',
            end: 29
          },
          CommentOperator {
            pos: 29,
            children: [],
            raw: <Buffer >,
            _raw: '',
            operator: VariableKey {
              pos: 32,
              children: [],
              raw: <Buffer 66 69>,
              _raw: 'fi',
              end: 34
            }
          }
        ],
        raw: <Buffer >,
        _raw: '',
        handleCheckCloseBlock: [Function: handleCheckCloseBlock],
        end: 34
      }
    },
    SpaceNewLine {
      pos: 34,
      children: [],
      raw: <Buffer 0a 0a>,
      _raw: '\\\\n\\\\n',
      end: 36
    },
    Variable {
      pos: 36,
      children: [
        VariableKey {
          pos: 36,
          children: [],
          raw: <Buffer 63 63 63>,
          _raw: 'ccc',
          end: 39
        },
        SymboEqual {
          pos: 39,
          children: [],
          raw: <Buffer 3d>,
          _raw: '=',
          end: 40
        },
        VariableValue {
          pos: 40,
          children: [],
          raw: <Buffer 64 64 64>,
          _raw: 'ddd'
        }
      ],
      raw: <Buffer >,
      _raw: '',
      keyVariable: VariableKey {
        pos: 36,
        children: [],
        raw: <Buffer 63 63 63>,
        _raw: 'ccc',
        end: 39
      },
      valueVariable: VariableValue {
        pos: 40,
        children: [],
        raw: <Buffer 64 64 64>,
        _raw: 'ddd'
      }
    }
  ],
  raw: <Buffer >,
  _raw: '',
  end: 43
}"
`;
